react-training/src/App.js

```diff
@@ -1,26 +1,189 @@
+                    prevTitle: "Card title5",
+                    prevText: "Text5"
+                },
+                {
+                    id: 6,
+                    showing: false,
+                    currentTitle: "Card title6",
+                    currentText: "Text6",
+                    titleCount: 0,
+                    textCount:0,
+                    prevTitle: "Card title6",
+                    prevText: "Text6"
+                },
+                {
+                    id: 7,
+                    showing: false,
+                    currentTitle: "Card title7",
+                    currentText: "Text7",
+                    titleCount: 0,
+                    textCount:0,
+                    prevTitle: "Card title7",
+                    prevText: "Text7"
+                },
+            ]
+        };
+
+        this.hideCheckBoxChange = this.hideCheckBoxChange.bind(this);
+        this.handleCheckboxChange = this.handleCheckboxChange.bind(this);
+        this.save = this.save.bind(this);
+        this.cancel = this.cancel.bind(this);
+        this.changeText = this.changeText.bind(this);
+        this.changeTitle = this.changeTitle.bind(this);
+        this.handleCheckbox = this.handleCheckbox.bind(this);
     }
 
     render() {
         return (
             <div className="container">
                 <div className="jumbotron">
                     <h1 className="display-3">App Name</h1>
-                </div>
-                <div>
-                    <Card state={this.state} th={this}></Card>
+                    <label><input type="checkbox" checked={this.state.isEditable} onChange={this.handleCheckbox}/>Read only</label>
+                    <CardList cards={this.state.cards} changeTitle={this.changeTitle} save={this.save}
+                        hideCheckBoxChange={this.hideCheckBoxChange} cancel={this.cancel} changeText={this.changeText}
+                        handleCheckboxChange={this.handleCheckboxChange} isEditable={this.state.isEditable}/>
                 </div>
             </div>
         );
     }
+
+    handleCheckboxChange(id) {
+        this.setState({cards: this.state.cards.map(card => {
+            if(card.id === id){
+                card.checked = !card.checked;
+            }
+            return card;
+        })})
+    }
+
+    handleCheckbox() {
+        this.setState({isEditable: !this.state.isEditable});
+        this.state.cards.map(card => {
+            if(card.showing === true){
+                this.cancel(card.id);
+            }
+        })
+    }
+
+    hideCheckBoxChange(id) {
+        this.setState({cards: this.state.cards.map(card => {
+            if(card.id === id){
+                card.showing = !card.showing;
+                card.checked = false;
+            }
+            return card;
+        })})
+    }
+
+    save(id) {
+        this.setState({cards: this.state.cards.map(card => {
+            if(card.id === id){
+                card.titleCount = 0;
```

Comment:
Could you please try to get rid of titleCount and textCount?
You can also get rid of prev/current title/text by putting the 'current' into the Card state